src/NHibernate/Async/Type/AbstractType.cs

```diff
@@ -0,0 +1,175 @@
+
+namespace NHibernate.Type
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public abstract partial class AbstractType : IType
+	{
+
+		/// <summary>
+		/// Reconstructs the object from its cached "disassembled" state.
+		/// </summary>
+		/// <param name="cached">The disassembled state from the cache</param>
+		/// <param name="session">The <see cref="ISessionImplementor"/> is not used by this method.</param>
+		/// <param name="owner">The parent Entity object is not used by this method</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		/// <returns>The assembled object.</returns>
+		/// <remarks>
+		/// This method calls DeepCopy if the value is not null.
+		/// </remarks>
+		public virtual Task<object> AssembleAsync(object cached, ISessionImplementor session, object owner, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			try
+			{
+				return Task.FromResult<object>(Assemble(cached, session, owner));
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<object>(ex);
+			}
+		}
+
+		public virtual Task BeforeAssembleAsync(object cached, ISessionImplementor session, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			try
+			{
+				BeforeAssemble(cached, session);
+				return Task.CompletedTask;
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<object>(ex);
+			}
+		}
+
+		/// <summary>
+		/// Retrieves an instance of the mapped class, or the identifier of an entity 
+		/// or collection from a <see cref="DbDataReader"/>.
+		/// </summary>
+		/// <param name="rs">The <see cref="DbDataReader"/> that contains the values.</param>
+		/// <param name="names">
+		/// The names of the columns in the <see cref="DbDataReader"/> that contain the 
+		/// value to populate the IType with.
+		/// </param>
+		/// <param name="session">the session</param>
+		/// <param name="owner">The parent Entity</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		/// <returns>An identifier or actual object mapped by this IType.</returns>
+		/// <remarks>
+		/// This method uses the <c>IType.NullSafeGet(DbDataReader, string[], ISessionImplementor, object)</c> method
+		/// to Hydrate this <see cref="AbstractType"/>.
+		/// </remarks>
+		public virtual Task<object> HydrateAsync(DbDataReader rs, string[] names, ISessionImplementor session, object owner, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			return NullSafeGetAsync(rs, names, session, owner, cancellationToken);
```

Comment:
> Does it makes sens to check CancellationRequested when otherwise it is just returning the task of another method?

In this particular case, it doesn't. [I will add](https://github.com/maca88/AsyncGenerator/issues/23) an option for fine tuning the generation for the method guards.